library/std/src/os/espidf/fs.rs

```diff
@@ -0,0 +1,126 @@
+                  methods of this trait"
+    )]
+    #[allow(deprecated)]
+    fn as_raw_stat(&self) -> &raw::stat;
+
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_dev(&self) -> u64;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_ino(&self) -> u64;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_mode(&self) -> u32;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_nlink(&self) -> u64;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_uid(&self) -> u32;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_gid(&self) -> u32;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_rdev(&self) -> u64;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_size(&self) -> u64;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_atime(&self) -> i64;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_atime_nsec(&self) -> i64;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_mtime(&self) -> i64;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_mtime_nsec(&self) -> i64;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_ctime(&self) -> i64;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_ctime_nsec(&self) -> i64;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_crtime(&self) -> i64;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_crtime_nsec(&self) -> i64;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_blksize(&self) -> u64;
+    #[stable(feature = "metadata_ext2", since = "1.8.0")]
+    fn st_blocks(&self) -> u64;
+}
+
+#[stable(feature = "metadata_ext", since = "1.1.0")]
+impl MetadataExt for Metadata {
+    #[allow(deprecated)]
+    fn as_raw_stat(&self) -> &raw::stat {
+        unsafe { &*(self.as_inner().as_inner() as *const libc::stat as *const raw::stat) }
+    }
+    fn st_dev(&self) -> u64 {
+        self.as_inner().as_inner().st_dev as u64
+    }
+    fn st_ino(&self) -> u64 {
+        self.as_inner().as_inner().st_ino as u64
+    }
+    fn st_mode(&self) -> u32 {
+        self.as_inner().as_inner().st_mode as u32
+    }
+    fn st_nlink(&self) -> u64 {
+        self.as_inner().as_inner().st_nlink as u64
+    }
+    fn st_uid(&self) -> u32 {
+        self.as_inner().as_inner().st_uid as u32
+    }
+    fn st_gid(&self) -> u32 {
+        self.as_inner().as_inner().st_gid as u32
+    }
+    fn st_rdev(&self) -> u64 {
+        self.as_inner().as_inner().st_rdev as u64
+    }
+    fn st_size(&self) -> u64 {
+        self.as_inner().as_inner().st_size as u64
+    }
+    fn st_atime(&self) -> i64 {
+        self.as_inner().as_inner().st_atime as i64
+    }
+    fn st_atime_nsec(&self) -> i64 {
+        //self.as_inner().as_inner().st_atime_nsec as i64
+        0
+    }
```

Comment:
Since this trait is platform-specific it doesn't need to have methods it can't implement properly.

Reply:
Agreed - will remove tomorrow as well. Ditto for all the others there returning 0.

Reply:
Btw, would you want me to continue force-pushing, or shall I start accumulating changes-after-feedback as separate commits, so that the progress is easier to track?

Reply:
either is fine. in the latter case you can rewrite the history before merge.

Reply:
OK. While we are at the `os/espidf/mod.rs` topic. Any hits regarding what shall I put as a "version" there?
I'm referring to this [FIXME](https://github.com/rust-lang/rust/pull/87666/files#diff-5b9c55eb49b42c03102371ec9b45d87ead04588d68415c07a1166186438a5eb9).

Reply:
Current nightly is 1.56, so if it were to land now and you wanted the new platform features to be instantly stable with the the first release you can put that version there.
If those traits are not critical for the platform to be usable and you have some reason to iterate on them you could mark them unstable instead.

Reply:
Oh wait, that's for the raw types, not the metadata extensions? I guess those are important and should be stable with the the first release.

Reply:
Neither is top priority, as using filesystems on the MCU is a bit of a niche case I think, due to the limited flash size in the MCU itself and the need to jump to external SD card for these use cases.

Let me sleep over it to figure out what is most appropriate regarding stable/unstable.
 
Threads and sockets is most important to get right first.

Reply:
OK, so I've force-pushed a cleanup of the `os/espidf` module:
* First of all, I decided to leave as stable everything, because, well, filesystems on the ESP-IDF might be a corner case but they are there, the `stat()` function works (even if a bit slow on the SPIFFS FS), so why shall I be seeking for unstable, where it does not make sense
* I've removed all commented out code that should not be there, double-checked that the `raw::stat` module is a faithful representation of what is in there in the ESP-IDF in terms of stat structure & members
* Some stubbed out code *MUST* stay and return 0. Turns out, the "platform specific" `MetadataExt` trait is not so platform specific, as it is used by `os::unix::fs` which is re-exposing its methods to the outside world. Since the ESP-IDF claims to be in the unix family, it *must* support all methods `os::unix::fs` re-expose, even if a subset of those (the `*_nsec` ones) return 0. I'm not setting a precedent: our sibling bare-metal RTOS `vxworks` port [does the same](https://github.com/rust-lang/rust/blob/master/library/std/src/os/vxworks/fs.rs#L82) with `*_nsec`, for the very same reasons.  
* Finally, I've kept versions at 1.1.0 (for the deprecated stuff) and at 1.8.0 for the `MetadataExt` trait as everybody else does. Again, an example: the vxworks port [was merged](https://github.com/rust-lang/rust/pull/61946) ~ 2019, while 1.8.0 is at least 3 years older (2016), let alone version 1.1.0. 
* Moreover: I would actually argue that re-using these versions as they are is even the correct behavior: imagine a cross-platform crate that uses this stuff, either via `os::unix::fs` or `os::platform` and sets the minimum compiler version requirement to 1.8.0. From the point of view of the crate that's the correct behavior, as everybody else except the ESP-IDF port is having those since 1.8.0. Now, people willing to consume that fictitious crate with ESP-IDF will of course never use a 1.8.0 version of the compiler simply because ESP-IDF support is not there. So putting a higher version on those crate items does not seem to help. Quite the opposite.

Reply:
@the8472 Just curious if you consider these changes acceptable?

Reply:
Sounds reasonable considering the precedent and that `os::unix` reexports stuff.